link/uprobe.go

```diff
@@ -0,0 +1,275 @@
+package link
+
+import (
+	"debug/elf"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal"
+	"github.com/cilium/ebpf/internal/unix"
+)
+
+func init() {
+	symbolsCache = internal.NewSymbolsCache()
```

Comment:
A global symbols cache is problematic: there is no way to release the memory used by obsolete entries without breaking the abstraction. Your motivation is that parsing symbols may be expensive?

I think I'd prefer to expose another type to the user, `type Executable struct` or something like that. In the constructor we do the parsing we need and store it in the type. After that the struct becomes read-only and is therefore safe to use concurrently / multiple times. Since it only contains data we don't need a `Close()` or similar, and the GC will remove it once unused. Having a separate type allows adding a constructor that takes detached debug info if necessary. `func Uprobe` would become `func (e *Executable) Uprobe(...)`, ditto for Uretprobe.

Reply:
I like the idea of using `Executable`! I'll ditch the cache

Reply:
Now I've thought it through a bit more, parsing an ELF into an `Executable` nicely addresses the most obvious problem with e.g. `Uprobe("/bin/bash", "readline", prog)`: it doesn't require re-parsing the binary when attaching multiple probes to the same binary. A use case for uprobes could be attaching a program as both a uprobe and uretprobe to the same symbol, or to multiple well-known symbols in the case of e.g. a Go or JVM runtime.

I like it!

Reply:
I tried to address all comments. All tests are still missing, I'd prefer to get a short review first